docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+# Record and Replay ADR
+
+## Submitters
+- Lenny Goodell (Intel)
+
+## Change Log
+- [Pending](https://github.com/edgexfoundry/edgex-docs/pull/863) (2022-09-20)
+
+## Referenced Use Case(s)
+- [Record and Replay UCR](https://docs.edgexfoundry.org/2.3/design/ucr/Record-and-Replay/)
+
+## Context
+This design involves creating a new Application Service that is responsible for the requirements in the above referenced UCR. This new Application Service will go above and beyond that of a normal Application Service,  thus it is architecturally significant.
+
+## Proposed Design
+A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. 
+
+!!! note
+    Application Services do not receive data when the Functions Pipelines are stopped.
+
+### Record Endpoint
+
+This `POST` API will start recording data as specified in the [request Data Transfer Object (DTO)](#record-request-dto) defined below. The request handler will validate the DTO and then create a new Functions Pipeline and [Start the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitrun) to process incoming data. 
+
+The Functions Pipeline will contain the following pipeline functions in the following order
+
+- [Filter functions](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#filtering) if filtering is needed, configured based on the DTO parameters.
+- [Batching pipeline function](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#batching) configured based on the DTO parameters. This will be used to control the record duration/count. 
+- New pipeline function written to process the batched data once the batching threshold has been exceeded. This function will simply send the recorded data to an async function for processing. 
+
+The async  function receiving the data will first [stop the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitstop) and then save the data for later replay and/or export. It will also determine the list of unique Device Profile and Device Names from the data and store them along side the recorded data. Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage. 
+
+!!! note
+    Starting a new recording will overwrite any previous recorded data.
```

Comment:
This implies that we don't wanna support parallel sessions, as opposed to what someone may expect from a web service. I would either allow parallel recordings or throw an error if a subsequent request is sent, forcing the user to DELETE before POSTing another. On the other hand, a PUT operation can be used to "replace" an existing recording.

Reply:
This not about multiple recording at once, it is about there only being one data set that will be overwritten.

> This implies that we don't wanna support parallel sessions

Correct, this first implementation is a crawl phase. Once we get usage and feedback we can add support for support parallel sessions.


> throw an error if a subsequent request is sent, forcing the user to DELETE before POSTing another

Yes, an error will be returned if a recording is already in progress. Will add wording for that (same for Replat) . The DELETE APIs to cancel a recording or replay have already been added.